sdk/eventhub/Azure.Messaging.EventHubs/stress/src/Infrastructure/EventTracking.cs

```diff
@@ -0,0 +1,246 @@
+    ///   intended partition was the one the event was received from, that the event was received in order, and that none were missed, and that
+    ///   the body is valid.
+    /// </summary>
+    ///
+    /// <param name="eventData"></param>
+    /// <param name="partitionReceivedFrom"></param>
+    /// <param name="metrics"></param>
+    ///
+    /// <returns>The publisher assigned index number of the event that was just read, or -1 if this event was unkown or a duplicate.</returns>
+    ///
+    /// <remarks>
+    ///   One instance should be created for each processor or consumer role. This instance will be able to keep track of all the events the
+    ///   processor or consumer has received for each partition it is reading or processing from.
+    /// </remarks>
+    ///
+    private int _checkEvent(EventData eventData, string partitionReceivedFrom, Metrics metrics)
+    {
+        // Id Checks
+        var hasID = eventData.Properties.TryGetValue(IdPropertyName, out var eventIdProperty);
+        var eventId = eventIdProperty?.ToString();
+
+        if (!hasID)
+        {
+            metrics.Client.GetMetric(Metrics.UnknownEventsProcessed).TrackValue(1);
+            return -1;
+        }
+
+        if (ReadEvents.ContainsKey(eventId))
+        {
+            metrics.Client.GetMetric(Metrics.DuplicateEventsDiscarded).TrackValue(1);
+            return -1;
+        }
+
+        // Partition Checks
+        var hasPartition = eventData.Properties.TryGetValue(PartitionPropertyName, out var partitionProperty);
+        var publisherSetPartitionProperty = partitionProperty?.ToString();;
+
+        if (partitionReceivedFrom != publisherSetPartitionProperty)
+        {
+            metrics.Client.GetMetric(Metrics.EventReceivedFromWrongPartition).TrackValue(1); //want an event here? or exception? This doesn't give much info
+        }
+
+        // Index Value checks
+        var hasIndex = eventData.Properties.TryGetValue(IndexNumberPropertyName, out var IndexProperty);
+        int.TryParse(IndexProperty.ToString(), out var indexNumber);
+
+        var seenEventFromThisPartition = LastReadPartitionSequence.TryGetValue(partitionReceivedFrom, out var lastReadFromPartition);
+
+        if (seenEventFromThisPartition && lastReadFromPartition >= indexNumber)
+        {
+            metrics.Client.GetMetric(Metrics.MissingOrOutOfOrderEvent).TrackValue(1);
+
+            var eventProperties = new Dictionary<string,string>();
+            eventProperties.Add(Metrics.PublisherAssignedIndex, indexNumber.ToString());
+            eventProperties.Add(Metrics.EventBody, eventData.EventBody.ToString());
+            metrics.Client.TrackEvent(Metrics.MissingOrOutOfOrderEvent, eventProperties);
+        }
+
+        LastReadPartitionSequence.AddOrUpdate(partitionReceivedFrom, _ => indexNumber, (k,v) => Math.Max(v, indexNumber));
+
+        // Hashed event body checks
+        using (SHA256 sha256Hash = SHA256.Create())
```

Comment:
I wonder if we would be better off making this class disposable and creating the hasher at the class level.   That would prevent us from having to allocate/destroy it multiple times.

Reply:
for that matter... making it `static` and just letting it live for the duration of the test run (and not bothering to dispose) is probably ok too...